Moonolith-based general interpolation between different meshes [moonolith-dev]

zulianp opened the pull request:
Interface between MFEM and the new Moonolith.
Volume transfer of discrete fields between arbitrarily distributed unstructured finite element meshes.

Superseeds #151.
<!--GHEX{"author":"zulianp","assignment":"2020-10-24T11:49:55","editor":"tzanio","id":315,"reviewers":["bslazarov","vladotomov","YohannDudouit","v-dobrev"],"merge":"2022-03-30T07:00:00","approval":"2022-03-23T16:35:46"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#315](https://github.com/mfem/mfem/pull/315) | @zulianp | @tzanio | @bslazarov + @vladotomov + @YohannDudouit + @v-dobrev | 10/24/20 | ⌛due 3/23/22 | ⌛due 3/30/22 | |
<!--ELBATXEHG-->

tzanio commented:
@zulianp will revive and prepare this branch for review.

tzanio commented:
Can you merge `master` in this branch and resolve the conflicts?

Once the Travis and Appveyor tests pass, mark the PR with the `ready-for-review` label.

tzanio commented:
Make sure you pull `master`. I do see the conflicts when I do the merge manually...

zulianp commented:
I see, my luck as always I merged just before it was updated.

tzanio commented:
LGTM

vladotomov commented on examples/CMakeLists.txt:
```diff
@@ -67,6 +67,9 @@ if (MFEM_USE_MPI)
     )
 endif()
 
+
```
Empty lines.

vladotomov commented on CMakeLists.txt:
```diff
@@ -335,6 +341,25 @@ if (MFEM_USE_PUMI)
   endif()
 endif()
 
+
```
Empty lines

vladotomov commented on CMakeLists.txt:
```diff
@@ -386,14 +412,17 @@ if (NOT DEFINED MFEM_TIMER_TYPE)
 endif()
 
 # List all possible libraries in order of dependencies.
+
```
empty lines

vladotomov commented on config/config.mk.in:
```diff
@@ -10,6 +10,7 @@
 # CONTRIBUTING.md for details.
 
 # Variables corresponding to defines in config.hpp (YES, NO, or value)
+
```
empty line

vladotomov commented on examples/moonolith/README.md:
```diff
@@ -0,0 +1,17 @@
+                               http://mfem.org
+
+This directory contains modifications of the example codes that illustrate the
+use of MFEM features based on ParMoonolith library for variation information transfer between non-matching meshes.
```
typo "variation"

vladotomov commented on makefile:
```diff
@@ -113,7 +113,9 @@ $(if $(word 2,$(SRC)),$(error Spaces in SRC = "$(SRC)" are not supported))
 
-EXAMPLE_SUBDIRS = amgx ginkgo hiop petsc pumi sundials superlu
+
+EXAMPLE_SUBDIRS = amgx ginkgo hiop petsc pumi sundials superlu moonolith
```
extra empty lines

zulianp replied:
Removed.

vladotomov commented on makefile:
```diff
@@ -400,8 +410,20 @@ ifneq (,$(filter install,$(MAKECMDGOALS)))
    export VERBOSE
 endif
 
+
```
empty lines

vladotomov commented on transfer/box.hpp:
```diff
@@ -0,0 +1,82 @@
+	void MaxCol(const DenseMatrix &mat, Vector &vec, bool include_vec_elements);
+	void MinCol(const DenseMatrix &mat, Vector &vec, bool include_vec_elements);
+
+	class Box {
```
Why is this functionality in MFEM and not in Moonolith?

zulianp replied:
Removed the class and used moonolith bounding volumes instead

vladotomov commented on transfer/hashgrid.hpp:
```diff
@@ -0,0 +1,39 @@
+
+
+namespace mfem {
+
```
It would be good to add a short explanation what these classes (all new classes) are used for. Also it would be good to explain why they must be in MFEM.

vladotomov replied:
ping @zulianp

zulianp replied:
I am currently away for the Swiss army COVID response. I will probably be back in January and I will have time then to look at the comments. Cheers!

tzanio replied:
take care Patrick!

zulianp replied:
It did not make sense to be in MFEM. The class has been removed.

vladotomov commented on CHANGELOG:
```diff
@@ -9,7 +9,9 @@
 Version 4.2.1 (development)
-===========================
+=========================================
+- Added support for parallel non-conforming, non-matching, variational, volumetric
```
What's the best place to download Moonolith?

vladotomov replied:
ping @zulianp

zulianp replied:
Added Bitbucket URL

vladotomov commented on examples/moonolith/ex1.cpp:
```diff
@@ -0,0 +1,159 @@
+
```
Could you add a short explanation what each example does + some sample runs.

zulianp replied:
Done.

zulianp commented:
Hi everyone! Sorry for taking so long. After I came back from deployment it took me a while to get over a large amount of work. I have tried to address all comments and improved the code.

Cheers

tzanio commented:
Welcome back Patrick!

We will get back to this after the mfem-4.3 release.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

vladotomov commented on config/defaults.mk:
```diff
@@ -353,6 +354,11 @@ ifeq ($(SLEPC_FOUND),YES)
 endif
 
+ifeq ($(MFEM_USE_MOONOLITH),YES)
+  include $(MOONOLITH_DIR)/config/moonolith-config.makefile
```
I cloned and installed moonolith, but I don't see this folder/file anywhere. Is this assuming something different than moonolith's `master` branch? It's not clear how to set `MOONOLITH_DIR`.

zulianp replied:
I have added more specific instructions. In short non-Cmake users need to install monolith with make install_all which will generate the make config file.

vladotomov commented on INSTALL:
```diff
@@ -508,6 +508,11 @@ MFEM_USE_MKL_CPARDISO = YES/NO
+MFEM_USE_MOONOOLITH = YES/NO
+   Enables the interface ParMoonolith for parallel non-conforming, non-matching,
+   variational, volumetric mesh information transfer using ParMoonolith. It
+   requires MFEM_USE_MPI=YES, MFEM_USE_METIS=YES, and MOONOLITH_DIR=<path to install>
```
It appears that you have implemented a serial `MortarAssembler` and examples that are serial. Are `MFEM_USE_MPI=YES` and metis strictly required for those?

zulianp replied:
Fixed it. Now it supports both MFEM_USE_MPI=YES and MFEM_USE_MPI=NO without any particular requirement.

vladotomov commented on examples/moonolith/ex2.cpp:
```diff
@@ -0,0 +1,158 @@
+//               domain methods for fluid-structure interaction or general
+//               multi-physics applications.
+//
+//               This particular example concerns vector FEM and it is only for serial runtimes.
```
The interface and code in `ex1` and `ex2` are very similar, maybe the two examples can be combined with a command line option for the space.

zulianp replied:
Done.

tzanio commented:
Please resolve conflicts.

zulianp commented:
Hello, Thanks for the feedback, I have addressed all comments. Please let me know if anything else is required.

Cheers

vladotomov commented on transfer/mortarassembler.hpp:
```diff
@@ -0,0 +1,73 @@
+   /*!
+    * @brief This method must be called before Assemble or Transfer.
+    * It will assemble the operator in all intersections found.
+    * @param integrator the integrator object
```
Again, some additional explanation how these integrators control the L2 projection, with respect to the cut elements,  would be great.

zulianp replied:
I have extended the explanation. I also added a link to the paper where everything is explained in detail.

vladotomov commented on examples/moonolith/ex1p.cpp:
```diff
@@ -0,0 +1,245 @@
+   }
+   else
+   {
+      src_fe_coll = make_shared<L2_FECollection>(1, src_mesh->Dimension());
```
The two FE orders should be a command line option.

zulianp replied:
Done.

zulianp commented on transfer/mortarassembler.hpp:
```diff
@@ -0,0 +1,73 @@
+    * @return true if there was an intersection and the operator has been
+    * assembled. False otherwise.
+    */
+   bool Assemble(std::shared_ptr<SparseMatrix> &B);
```
Done. Now Init() and Apply(...) can be used.

zulianp commented on transfer/mortarintegrator.hpp:
```diff
@@ -0,0 +1,118 @@
+public:
+   VectorL2MortarIntegrator() { Init(NULL, NULL, NULL); }
+   VectorL2MortarIntegrator(Coefficient *_q) { Init(_q, NULL, NULL); }
+   VectorL2MortarIntegrator(Coefficient &q) { Init(&q, NULL, NULL); }
```
Done.

zulianp commented on transfer/mortarintegrator.hpp:
```diff
@@ -0,0 +1,118 @@
+{
+
+/*!
+ * @brief Interface for mortar element assembly
```
Added explanation.  This functionality is required by the moonolith integration, but it can be used also without moonolith.

zulianp commented on transfer/parallel/pmortarassembler.cpp:
```diff
@@ -0,0 +1,1098 @@
+   predicate->add(0, 1);
+
+   MOONOLITH_EVENT_BEGIN("create_adapters");
+   ////////////////////////////////////////////////////////////////////////////////////////////////////
```
:). Done

zulianp commented on transfer/mortarassembler.hpp:
```diff
@@ -0,0 +1,73 @@
+#ifndef MFEML2P_MORTAR_ASSEMBLER_HPP
+#define MFEML2P_MORTAR_ASSEMBLER_HPP
+
```
Added headers

zulianp commented on mfem.hpp:
```diff
@@ -43,4 +43,8 @@
 #include "fem/fem.hpp"
 
+#ifdef MFEM_USE_MOONOLITH
+#include "transfer/transfer.hpp"
```
Done.

zulianp commented on mesh/mesh.hpp:
```diff
@@ -17,6 +17,8 @@
 #include "../general/globals.hpp"
 #include "triangle.hpp"
 #include "tetrahedron.hpp"
+#include "quadrilateral.hpp" //FIXME maybe there was a reason why these includes were not here
```
Removed

zulianp commented on examples/moonolith/ex1p.cpp:
```diff
@@ -0,0 +1,245 @@
+// Moonolith sample runs:
+//               mpirun -np 4 ex1p
+//               mpirun -np 4 ex1p --source_refinements 1 --dest_refinements 2
+//               mpirun -np 4 ex1p --source_refinements 1 --dest_refinements 2 --use_vector_fe
```
We are aware of the issue with RT,  I described it as an experimental feature in the doc. The problem appears only in parallel runs... maybe I am missing some permutation since the sum of the operator seems to be consistent with serial runs. Any help in this matter would be appretiated.

zulianp commented on examples/moonolith/ex2p.cpp:
```diff
@@ -0,0 +1,177 @@
+// Moonolith sample runs:
+//               mpirun -np 4 ex2p
+//               mpirun -np 4 ex2p --source_refinements 1 --dest_refinements 2
+//               mpirun -np 4 ex2p -s ../../data/inline-hex.mesh -d ../../data/inline-tet.mesh
```
Fixed! Sorry, I forgot to test the new version of the code with simplicial meshes.

zulianp commented:
I have addressed most comments with the exception of the parallel RT-FE transfer (see https://github.com/mfem/mfem/pull/315#discussion_r681619773). The main use of this library is with the L2_FECollection, but it would be nice if we could make it work in parallel also for the RT_FECollection. Worst case it can be limited to serial runs for the time being with a failure.

vladotomov commented on CHANGELOG:
```diff
@@ -10,7 +10,10 @@
+- Added support for parallel non-conforming, non-matching, variational, volumetric
+  mesh information transfer using ParMoonolith (https://bitbucket.org/zulianp/par_moonolith).
+  With ParMortarAssember fields can be exchanged between arbitrarily distributed and unrelated
+  finite element meshes in a variationally consistent way.
```
How do you expect it to work with curved meshes? Is this considered in the intersection method? I tried one run, but couldn't get it working with the current setup:
`./ex1p -s ../../miniapps/gslib/triple-pt-1.mesh -dr 3`
I also uncommented `x_new *= 7.0` (line 30) to cover the domain.

<img width="1289" alt="Screen Shot 2021-08-03 at 5 18 40 PM" src="https://user-images.githubusercontent.com/17843342/128102518-b2923b1e-cb15-406e-87a7-01d23975c12b.png">

zulianp replied:
Curved meshes are not supported at this time.  I have added a comment in the documentation. 
If there is strong interest in the future I could add approximate quadrature routines for warped meshes.

vladotomov reviewed (approved):
Very nice work @zulianp, this is a great capability that can be useful in many ways. Here are two things I thought of:

- For transfers on curved meshes, one could think about combining this with the LOR capabilities.
- It would be great to have a method that simply integrates a FE function on a sub-region, especially if the sub-region can be specified by a FE level set function.

zulianp commented:
Thanks, I enjoyed the process. We can have eventually a chat about more generalized integration. I will be writing a proposal in September for FSI algorithms, maybe I can fit something in there at least at the algorithmic level, which can then be adapted to MFEM.

tzanio commented:
@rw-anderson and @v-dobrev -- will you like to review this?

v-dobrev commented:
> @rw-anderson and @v-dobrev -- will you like to review this?

We do need 1 more review. I'll need more time to review the PR.

YohannDudouit commented on fem/moonolith/parallel/pmortarassembler.hpp:
```diff
@@ -0,0 +1,118 @@
+    * element space
+    * @return true if there was an intersection and the output can be used.
+    */
+   bool Transfer(ParGridFunction &src_fun, ParGridFunction &dest_fun);
```
```suggestion
   bool Transfer(const ParGridFunction &src_fun, ParGridFunction &dest_fun);
```
?

YohannDudouit commented on fem/moonolith/parallel/pmortarassembler.hpp:
```diff
@@ -0,0 +1,118 @@
+    * element space
+    * @return true if the transfer was succesfull, fale otherwise.
+    */
+   bool Apply(ParGridFunction &src_fun, ParGridFunction &dest_fun);
```
```suggestion
   bool Apply(const ParGridFunction &src_fun, ParGridFunction &dest_fun);
```
?

YohannDudouit commented on fem/moonolith/parallel/pmortarassembler.hpp:
```diff
@@ -0,0 +1,118 @@
+    * @return true if there was an intersection and the operator has been
+    * assembled. False otherwise.
+    */
+   bool Init();
```
ditto, find a better name

YohannDudouit commented on fem/moonolith/transfer.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+#endif
+
+} // namespace mfem
```
newline

YohannDudouit commented on fem/moonolith/transfer.hpp:
```diff
@@ -0,0 +1,36 @@
+namespace mfem
+{
+
+void InitTransfer(int argc, char *argv[]);
+int FinalizeTransfer();
+
+#ifdef MFEM_USE_MPI
+void InitTransfer(int argc, char *argv[], MPI_Comm comm);
+#endif
```
Documentation?

YohannDudouit commented on fem/moonolith/transferutils.cpp:
```diff
@@ -0,0 +1,156 @@
+namespace mfem
+{
+
+namespace private_
```
We usually use `internal`, would that work in your case too?

YohannDudouit commented on fem/moonolith/transferutils.hpp:
```diff
@@ -0,0 +1,39 @@
+namespace private_
+{
+
+Element *NewElem(const int type, const int *cells_data, const int attr);
+void Finalize(Mesh &mesh, const bool generate_edges);
+
+void MaxCol(const DenseMatrix &mat, double *vec, bool include_vec_elements);
+void MinCol(const DenseMatrix &mat, double *vec, bool include_vec_elements);
+
+int MaxVertsXFace(const int type);
+double Sum(const DenseMatrix &mat);
```
Minimal documentation?

YohannDudouit reviewed (commented):
Thanks @zulianp for this PR, here are my first remarks on the style/design.

zulianp commented:
Thanks @YohannDudouit for your review, I addressed all comments and applied all the suggested changes :-)

tzanio commented:
ping: @YohannDudouit

YohannDudouit commented:
@zulianp  My legendary `cmake` skills have hit again, I did:
```
git clone https://bitbucket.org/zulianp/moonolith.git
cd moonolith
mkdir build
cd build
cmake ..
```
and I get the following error:
```
 ~/Projects/moonolith/build/> cmake ..
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is AppleClang 11.0.0.11000033
-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- [Status] MOONOLITH_ROOT_PATH: /Users/dudouit1/Projects/moonolith/src
-- Found Doxygen: /usr/local/bin/doxygen (found version "1.9.3") found components: doxygen missing components: dot
CMake Error at cmake/FindNETCDF.cmake:176 (else):
  Flow control statements are not properly nested.
Call Stack (most recent call first):
  src/io/CMakeLists.txt:27 (find_package)


CMake Warning at src/io/CMakeLists.txt:27 (find_package):
  By not providing "FindNETCDF.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "NETCDF", but
  CMake did not find one.

  Could not find a package configuration file provided by "NETCDF" with any
  of the following names:

    NETCDFConfig.cmake
    netcdf-config.cmake

  Add the installation prefix of "NETCDF" to CMAKE_PREFIX_PATH or set
  "NETCDF_DIR" to a directory containing one of the above files.  If "NETCDF"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
```
Do you know if the problem is on my side or on your side?

zulianp commented:
@YohannDudouit Use this code repository instead https://bitbucket.org/zulianp/par_moonolith :-)

YohannDudouit commented on INSTALL:
```diff
@@ -520,6 +520,16 @@ MFEM_USE_MKL_CPARDISO = YES/NO
    MFEM_USE_LAPACK=YES, verify that the MKL LAPACK libraries are used. The
    OpenMP capabilities are disabled at link time.
 
+MFEM_USE_MOONOOLITH = YES/NO
```
I think there is a typo here.
```suggestion
MFEM_USE_MOONOLITH = YES/NO
```

YohannDudouit commented:
@zulianp Am I doing something wrong?
```
 ~/Projects/mfem/> make pdebug MFEM_USE_MOONOLITH=YES MOONOLITH_PATH=.. -j16
config/defaults.mk:373: /config/moonolith-config.makefile: No such file or directory
make: *** No rule to make target `/config/moonolith-config.makefile'.  Stop.
```

zulianp commented:
> @zulianp Am I doing something wrong?
> 
> ```
>  ~/Projects/mfem/> make pdebug MFEM_USE_MOONOLITH=YES MOONOLITH_PATH=.. -j16
> config/defaults.mk:373: /config/moonolith-config.makefile: No such file or directory
> make: *** No rule to make target `/config/moonolith-config.makefile'.  Stop.
> ```

@YohannDudouit 

There is a section dedicated to Makefile users here
https://bitbucket.org/zulianp/par_moonolith/src/master/

In short: use `make install_all` for installing also the moonolith-config.makefile.

Then set the to the directory where you installed the library 
MOONOLITH_DIR instead of MOONOLITH_PATH

YohannDudouit commented:
@zulianp Do you think you can do something about these warnings?
```
In file included from fem/moonolith/parallel/pmortarassembler.cpp:22:
/usr/local/include/moonolith_redistribute.hpp:140:27: warning: unused variable 'n_incoming' [-Wunused-variable]
            const Integer n_incoming = comm_.i_send_recv_all(send_buffers, recv_buffers, comm_tag);
                          ^
fem/moonolith/parallel/pmortarassembler.cpp:413:13: warning: unused variable 'elem' [-Wunused-variable]
      auto &elem = *space.GetFE(k);
            ^
fem/moonolith/parallel/pmortarassembler.cpp:554:18: warning: unused variable 'global_id' [-Wunused-variable]
      int index, global_id;
                 ^
fem/moonolith/parallel/pmortarassembler.cpp:797:8: warning: unused variable 'skip_zeros' [-Wunused-variable]
   int skip_zeros = 1;
       ^
fem/moonolith/parallel/pmortarassembler.cpp:824:19: warning: unused variable 'src_mesh' [-Wunused-variable]
      const auto &src_mesh = *src.GetMesh();
                  ^
fem/moonolith/parallel/pmortarassembler.cpp:983:20: note: in instantiation of function template specialization 'mfem::Assemble<2>' requested here
      return mfem::Assemble<2>(comm, impl_->source, impl_->destination,
                   ^
fem/moonolith/parallel/pmortarassembler.cpp:825:19: warning: unused variable 'dest_mesh' [-Wunused-variable]
      const auto &dest_mesh = *dest.GetMesh();
                  ^
fem/moonolith/parallel/pmortarassembler.cpp:824:19: warning: unused variable 'src_mesh' [-Wunused-variable]
      const auto &src_mesh = *src.GetMesh();
                  ^
fem/moonolith/parallel/pmortarassembler.cpp:990:20: note: in instantiation of function template specialization 'mfem::Assemble<3>' requested here
      return mfem::Assemble<3>(comm, impl_->source, impl_->destination,
                   ^
fem/moonolith/parallel/pmortarassembler.cpp:825:19: warning: unused variable 'dest_mesh' [-Wunused-variable]
      const auto &dest_mesh = *dest.GetMesh();
                  ^
8 warnings generated.
```

YohannDudouit commented:
@zulianp This looks suspicious to me, is it expected?
![Screen Shot 2022-02-02 at 4 16 07 PM](https://user-images.githubusercontent.com/190171/152259581-5ed5cfdb-820c-4ef7-94eb-b8e8bcb42eb5.png)

YohannDudouit commented:
@zulianp Increasing the destination order makes it worse it seems:
![Screen Shot 2022-02-02 at 4 19 57 PM](https://user-images.githubusercontent.com/190171/152259969-2f036008-fd51-42e4-901c-0a436b51202c.png)
:

zulianp commented:
> @zulianp This looks suspicious to me, is it expected? ![Screen Shot 2022-02-02 at 4 16 07 PM](https://user-images.githubusercontent.com/190171/152259581-5ed5cfdb-820c-4ef7-94eb-b8e8bcb42eb5.png)

@YohannDudouit The elements of the mesh are not affine, maybe that is the reason.  But in principle, it is not supposed to happen. Can you give me the unstructured mesh?

YohannDudouit commented:
> 

@zulianp There is extra meshes in the `mfem/data`, here is the unstructured mesh: https://github.com/mfem/data/blob/master/gmsh/v22/unstructured_quad.v22.msh

zulianp commented:
@YohannDudouit  It will take some time to figure out why these examples are not working (several versions ago... I am sure I did not encounter such issues even for examples like these, so I might have broken something).

zulianp commented:
@YohannDudouit  the issue is resolved, now it should be more robust for these kinds of examples

![Screenshot 2022-02-11 at 14 12 02](https://user-images.githubusercontent.com/11992284/153597676-4817450f-89a9-4af8-ae76-3bb963340263.png)

YohannDudouit commented:
@zulianp It looks like there still are problems when increasing the destination order:
![Screen Shot 2022-02-15 at 9 59 20 AM](https://user-images.githubusercontent.com/190171/154121373-7230f59e-a781-4d29-9b1f-0aa99a76b095.png)

zulianp commented:
> @zulianp You missed a warning:
> 
> ```
> /usr/local/include/moonolith_redistribute.hpp:140:27: warning: unused variable 'n_incoming' [-Wunused-variable]
>             const Integer n_incoming = comm_.i_send_recv_all(send_buffers, recv_buffers, comm_tag);
>                           ^
> 1 warning generated.
> ```

Fixed it on the moonolith repository

zulianp commented:
> @zulianp It looks like there still are problems when increasing the destination order: ![Screen Shot 2022-02-15 at 9 59 20 AM](https://user-images.githubusercontent.com/190171/154121373-7230f59e-a781-4d29-9b1f-0aa99a76b095.png)



> @zulianp It looks like there still are problems when increasing the destination order: ![Screen Shot 2022-02-15 at 9 59 20 AM](https://user-images.githubusercontent.com/190171/154121373-7230f59e-a781-4d29-9b1f-0aa99a76b095.png)

Ouch. It seems that the problem is not the mesh though but higher order. I tried also the standard quad mesh I get the same problem. I will take me some time to guess what went wrong with the changes.

zulianp commented:
@YohannDudouit it was good that you tried with that mesh, now everything is definitely better. Here is a summary of the changes

1) Variational transfer extended to non-affine elements with linear faces.
   - When performing quadrature on elements with nonlinear transformation, small local numerical errors create unwanted mass contributions in the coupling operator. The problem is typically removed if also the mass matrix uses the same quadrature points.
2) More control over the number of iterations of mass matrix inversion.
3) Fixed quadrature order, which for non-simplicial elements would be underestimated in certain cases.



Works up to destination order 8:

![Screenshot 2022-02-17 at 15 13 32](https://user-images.githubusercontent.com/11992284/154501890-d41f9f79-8af6-4a3a-baed-500814730364.png)

YohannDudouit commented:
@zulianp I think I found a last bug in `ex2p` in 3D:
![Screen Shot 2022-03-07 at 2 35 12 PM](https://user-images.githubusercontent.com/190171/157129580-4c74b66b-ffd0-466f-aab3-2bb4ba5ffc69.png)

YohannDudouit commented:
@zulianp The error seems very localized:
![Screen Shot 2022-03-07 at 2 59 51 PM](https://user-images.githubusercontent.com/190171/157132561-59c37b29-1a20-41d1-9953-b2dd75cbd6c7.png)

zulianp commented:
@YohannDudouit , That looks like a tough mesh :-). But a very good robustness test. I will work on it in the next few weeks

YohannDudouit commented:
@zulianp I saw you pushed some commits? Is it good now?

zulianp commented:
@YohannDudouit, Yes.  I believe this example you ran, is the limit case for my code.  As you can see in the picture, it is able to handle it now. But other variants of this technique might be required in the future in order to graciously handle such non-standard scenarios. I extended the documentation to include these limitations.

<img width="1238" alt="Screenshot 2022-03-22 at 08 55 59" src="https://user-images.githubusercontent.com/11992284/159434753-859eecee-c2ba-457d-a6ea-c27ce42eedeb.png">
I

YohannDudouit reviewed (approved):
Thanks @zulianp for adding this!

zulianp commented:
> Thanks @zulianp for adding this!

Thank you for taking the time for reviewing :-)

tzanio commented:
Merged in `next` for testing...

tzanio reviewed (commented):
Can you please fully update the INSTALL file (see e.g. CALIPER)

tzanio commented:
@zulianp -- I made a few small adjustments in https://github.com/mfem/mfem/pull/315/commits/b71568dac9d941a22e5b6ce11801c4c03351475e, can you take a look and check you are OK with them?

Also -- any chance we can remove the `fem/moonolith/parallel` directory and just have the `pmortarassembler` files in `fem/moonolith`?

tzanio commented:
Thank you @zulianp. If all nightly tests pass, I will merge it tomorrow.

zulianp commented:
Thanks to you @tzanio.  :-)

tzanio reviewed (approved):
Thank for this really nice contributions @zulianp, and for your patience and cooperation in getting it merged. 🙏 

We have to advertise it, but I think this will be really useful for many of our applications. 👍 

Congratulations 🎉

zulianp commented:
Thanks! It was my pleasure, I enjoyed the process and learned much.  :-)

I am available for advertising and understanding if and what eventual extensions might be interesting or useful to a wider audience.
